app/notification.go

```diff
@@ -578,47 +610,67 @@ func getMentionsEnabledFields(post *model.Post) model.StringArray {
+func (a *App) getMentionKeywordsInChannel(profiles map[string]*model.User, allowChannelMentions bool, channelMemberNotifyPropsMap map[string]model.StringMap) map[string][]string {
 	keywords := make(map[string][]string)
 
-	for id, profile := range profiles {
-		userMention := "@" + strings.ToLower(profile.Username)
-		keywords[userMention] = append(keywords[userMention], id)
-
-		if len(profile.NotifyProps[model.MENTION_KEYS_NOTIFY_PROP]) > 0 {
-			// Add all the user's mention keys
-			splitKeys := strings.Split(profile.NotifyProps[model.MENTION_KEYS_NOTIFY_PROP], ",")
-			for _, k := range splitKeys {
-				// note that these are made lower case so that we can do a case insensitive check for them
-				key := strings.ToLower(k)
-				keywords[key] = append(keywords[key], id)
-			}
-		}
+	for _, profile := range profiles {
+		addMentionKeywordsForUser(
+			keywords,
+			profile,
+			channelMemberNotifyPropsMap[profile.Id],
+			GetStatusFromCache(profile.Id),
+			allowChannelMentions,
+		)
+	}
 
-		// If turned on, add the user's case sensitive first name
-		if profile.NotifyProps[model.FIRST_NAME_NOTIFY_PROP] == "true" {
-			keywords[profile.FirstName] = append(keywords[profile.FirstName], profile.Id)
-		}
+	return keywords
+}
 
-		ignoreChannelMentions := false
-		if ignoreChannelMentionsNotifyProp, ok := channelMemberNotifyPropsMap[profile.Id][model.IGNORE_CHANNEL_MENTIONS_NOTIFY_PROP]; ok {
-			if ignoreChannelMentionsNotifyProp == model.IGNORE_CHANNEL_MENTIONS_ON {
-				ignoreChannelMentions = true
-			}
+// addMentionKeywordsForUser adds the mention keywords for a given user to the given keyword map. Returns the provided keyword map.
+func addMentionKeywordsForUser(keywords map[string][]string, profile *model.User, channelNotifyProps map[string]string, status *model.Status, allowChannelMentions bool) map[string][]string {
+	userMention := "@" + strings.ToLower(profile.Username)
+	keywords[userMention] = append(keywords[userMention], profile.Id)
+
+	if len(profile.NotifyProps[model.MENTION_KEYS_NOTIFY_PROP]) > 0 {
+		// Add all the user's mention keys
+		splitKeys := strings.Split(profile.NotifyProps[model.MENTION_KEYS_NOTIFY_PROP], ",")
+		for _, k := range splitKeys {
+			// note that these are made lower case so that we can do a case insensitive check for them
+			key := strings.ToLower(k)
+			keywords[key] = append(keywords[key], profile.Id)
 		}
+	}
 
-		// Add @channel and @all to keywords if user has them turned on
-		if lookForSpecialMentions {
-			if int64(len(profiles)) <= *a.Config().TeamSettings.MaxNotificationsPerChannel && profile.NotifyProps[model.CHANNEL_MENTIONS_NOTIFY_PROP] == "true" && !ignoreChannelMentions {
-				keywords["@channel"] = append(keywords["@channel"], profile.Id)
-				keywords["@all"] = append(keywords["@all"], profile.Id)
+	// If turned on, add the user's case sensitive first name
+	if profile.NotifyProps[model.FIRST_NAME_NOTIFY_PROP] == "true" {
```

Comment:
since this is a special case, shouldn't we store it separately? later on, we are going to check it with all the case insensitive.
this also would allow us to detect if we shouldn't run a second time through case sensitive words if there are none, reducing iterations.

Reply:
I'll check to see if there's a general one for improving this area and add this to it if it exists. If I can't find one, I'll make a separate ticket for this since it would be a decent performance improvement

Reply:
https://mattermost.atlassian.net/browse/MM-18477